Add search pills

evykassirer opened the pull request:
This PR is ready for review. Remaining tasks are being tracked below.

Progress towards issue #27150. See #25669 for previous (and incomplete) implementation that was removed.

---

Issues found in test deploy:

* [x] Opening the search bar, allowing the current pills to be cleared easily -- [CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20pills.20.2326803.3A.20new.20search.20behavior/near/1837042)
* [x] When I'm typing a search term and haven't yet turned it into a pill, if I hit the left-arrow key, it seems to take me out of the search box entirely and move the focus to somewhere else on the page. (Like somewhere in the middle of the recent-conversations table, if that's the screen I'm on.) --> only in recent conversations
* [x] When you load into a narrow with [near](https://chat.zulip.org/#narrow/stream/101-design/topic/Implement.203-way.20compose.20box.20resizing.20.2329966/near/1832732) in the URL, it loads the page with focus on search bar
* [x] cursor in the search box is too high all the time on Chrome and Safari
* [x] is:followed isn't supported
* [x] commas in topic names cause issues [CZO thread](https://chat.zulip.org/#narrow/stream/6-frontend/topic/search.20pills.20comma.20split/near/1836848)
* [x] Pressing Enter when there's a "search for..." term should just submit any time a search term is at the end of the search box


Open questions/issues that I'm looking for help with


* [ ] Should I test `is_using_input_method` use cases? I don't know much about it and haven't prioritized it yet.

Things that aren't active todos for me, but worth nothing:

* I added `input_pill` and `search_suggestion` to `EXEMPT_FILES` from codecov. I'm open to adding more test coverage for that code, but only after code review, since it's often a pain to write these kinds of tests and I want some confirmation that the implementation I'd be testing is what we want.

Out of scope for this PR (unless people want otherwise):
* Most of [#29510](https://github.com/zulip/zulip/issues/29510)
* Editing previous group user pills
* More fine-grained keyboard navigation for group pills
* Excluded words and messaging around it ([CZO thread](https://chat.zulip.org/#narrow/stream/9-issues/topic/Quoted.20common.20words.20are.20excluded.20but.20do.20not.20say.20so/near/1759248))
* Backspace for pill deletion: We'd change backspace behavior for pills to select the pill on the first backspace and delete the whole pill on the second backspace. If the pill is already highlighted from left-pressing, then backspace would delete it right away.
* Make pills copyable. Right now there's e.g. no way to easily copy the list of recipients in a DM recipient box.

evykassirer commented on web/src/message_view_header.js:
```diff
@@ -2,13 +2,18 @@ import $ from "jquery";
+import * as search from "./search";
+import * as search_pill from "./search_pill";
+
+// put widget in its own file to avoid circular imports?
```
Or maybe... we move all the search related code in `message_view_header.js` to `search.js`, such that `message_view_header` can import `search` instead of vice versa.

https://chat.zulip.org/#narrow/stream/6-frontend/topic/search.20code.20files

evykassirer commented on web/src/message_view_header.js:
```diff
@@ -191,6 +195,9 @@ export function open_search_bar_and_close_narrow_description() {
     if ($("#search_query").text() === "") {
+        // TODO: this is happening too late, or at least the typeahead
+        // `source` happens before this, so it doesn't exclude the
+        // right things from the dropdown options.
```
I think this might be a consequence of an existing bug in the redesign actually https://github.com/zulip/zulip/pull/24345#issuecomment-1729306042

evykassirer commented on web/src/search.js:
```diff
@@ -6,12 +6,14 @@ import {Filter} from "./filter";
+export let search_pill_widget = null;
 
-export function narrow_or_search_for_term(search_string, {on_narrow_search}) {
+function narrow_or_search_for_term({on_narrow_search}) {
```
The `export` already wasn't needed, I can make a separate commit for removing it.

evykassirer commented on web/styles/search.css:
```diff
@@ -244,16 +248,16 @@
+            /* Add line-height equal to height to mimic baseline alignment. */
+            line-height: 20px;
+            align-self: center;
+        }
```
Scoped these as styling only for pills within the typeahead -- can make this a separate commit.

evykassirer commented on web/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -175,9 +175,6 @@ import {get_string_diff} from "../../src/util";
     }
-    // The naturalSearch option causes arrow keys to immediately
-    // update the search box with the underlying values from the
-    // search suggestions.
```
This isn't relevant anymore with pills.

timabbott replied:
Yeah probably making this a follow-up commit would simplify review; should be very easy to split with `git checkout` to make removing this the first commit, and then reordering it to the end.

timabbott commented on web/src/message_view_header.js:
```diff
@@ -93,12 +93,16 @@ function append_and_display_title_area(message_view_header_data) {
 
+function set_search_bar_text(text) {
+    $("#search_query").val(text);
+}
```
For this sort of prep commit, it'd good to have an explanation in the commit message for why this is a good interface to extract -- e.g. some note about how the logic for setting the search bar text will get more complicated.

Ayush2k02 commented:
@evykassirer have you set up the typeahead source for the pill in searchpill.js

evykassirer commented on web/src/search.js:
```diff
@@ -131,16 +152,7 @@ export function initialize({on_narrow_search}) {
-                // this codepath is that they first all blur the box to
-                // indicate that they've done what they need to do)
-
-                // Pill is already added during keydown event of input pills.
```
I don't remember why I deleted this comment, but I'm also not sure what situations this actually happens, so I'll look into it. Might be able to delete it, which would be great.

evykassirer commented on web/src/search.js:
```diff
@@ -131,16 +152,7 @@ export function initialize({on_narrow_search}) {
 
             if (keydown_util.is_enter_event(e) && $search_query_box.is(":focus")) {
-                // We just pressed Enter and the box had focus, which
-                // means we didn't use the typeahead at all.  In that
```
This isn't true anymore -- using enter in the typeahead causes pills to be created. You have to press enter again when the box has focus to trigger a search

Ayush2k02 commented:
just a quick question, @evykassirer. When a user searches for a stream, would the search string be converted to a pill or the corresponding stream which matched, would convert to the pill ?

evykassirer commented:
@Ayush2k02 I don't think I understand your question, can you explain more? Is those two possibiliites a difference in user experience or a technical implementation difference?

evykassirer commented:
Yes that's what would happen. You can also play around with this PR locally if you want to see more how it works: `./tools/fetch-pull-request 26803`

Ayush2k02 commented:
Thanks for this tip, @evykassirer I was actually trying to integrate the search_pill with something, this would solve a lot of headache for me.

Ayush2k02 commented:
@evykassirer are planning to migrate search_pill to typescript ?

evykassirer commented:
That's part of the plan going forward after this first PR, but I haven't looked much yet into what it would involve.

Ayush2k02 commented:
@evykassirer , have you converted search_pill to typescript yet, below is my implimentation.
`import {Filter} from "./filter";
import type {InputPillContainer, InputPillItem} from "./input_pill";
import * as input_pill from "./input_pill";

type SearchPill = {
    description_html : string;
};
export type SearchPillWidget = InputPillContainer<SearchPill>;

export function create_item_from_search_string(search_string : string) : InputPillItem<SearchPill>{
    const operator = Filter.parse(search_string);
    const description_html = Filter.search_description_as_html(operator);
    return {
        type: "search",
        display_value: search_string,
        description_html,
    };
}

export function get_search_string_from_item(item: InputPillItem<SearchPill>) : string {
    return item.display_value;
}

export function create_pills($pill_container: JQuery) : input_pill.InputPillContainer<SearchPill> {
    const pills = input_pill.create({
        $container: $pill_container,
        create_item_from_text: create_item_from_search_string,
        get_text_from_item: get_search_string_from_item,
    });
    return pills;
}

export function append_search_string(search_string : string, pill_widget : SearchPillWidget) : void {
    const items = pill_widget.items();
    const search_strings = new Set(items.map((item) => item.display_value));

    const operators = Filter.parse(search_string);
    for (const operator of operators) {
        const input = Filter.unparse([operator]);
        if (!search_strings.has(input)) {
            pill_widget.appendValue(input);
        }
    }
    pill_widget.clear_text();
}

export function get_search_string_for_current_filter(pill_widget : SearchPillWidget) : string {
    const items = pill_widget.items();
    const search_strings = items.map((item) => item.display_value);
    return search_strings.join(" ");
}`

evykassirer commented:
I've updated this PR to have the search pill file be in typescript

timabbott commented:
This needs a rebase following https://github.com/zulip/zulip/pull/28054.

timabbott commented on web/src/search_pill.ts:
```diff
@@ -0,0 +1,49 @@
+import * as input_pill from "./input_pill";
+import type {InputPillContainer} from "./input_pill";
+
+type SearchPill = {display_value: string; type: string; description_html: string};
```
I would wrap this with each value on its own line for better readability.

evykassirer replied:
Oh weird yeah I expected the linter to do that. Fixed

timabbott commented on web/src/search_pill.ts:
```diff
@@ -0,0 +1,49 @@
+
+export function append_search_string(search_string: string, pill_widget: SearchPillWidget): void {
+    const items = pill_widget.items();
+    const search_strings = new Set(items.map((item) => item.display_value));
```
Would `existing_search_strings` be a clearer variable name here?

This comment might also benefit from a docstring style comment.

evykassirer replied:
(no longer relevant, see below comment)

timabbott commented on web/src/search.js:
```diff
@@ -58,25 +67,47 @@ export function initialize({on_narrow_search}) {
+            if (query !== "") {
+                search_input_has_changed = true;
+            }
             const suggestions = search_suggestion.get_suggestions(query);
```
Hmm. Structurally, won't `get_suggestions` want to have the existing pills passed as a parameter, so it can do things like "Don't offer DM options if you've already got a `stream` pill"? 

Even if we don't consider it a requirement to implement a full set of such logic in a first version, and thus `search_suggestion.js` just ignores the value, I think that's the right interface here, rather than doing after-the-fact duplicate filtering in this function?

evykassirer replied:
Yeah, as I just noted in another comment, I've been working on moving to that strategy. Good callout!

timabbott commented:
I didn't see anything that needs a major rework in this initial read, though I've not tested manually pending a rebase.

Posted a bunch of smaller comments on this that should help clean up/clarify the implementation. I think a good next step is to rebase this, addressing those tactical comments and do some intensive local QA on this, do a round of improvements based on that, and then try deploying on chat.zulip.org.

I would be a bit surprised if we are able to complete something here without making any changes to `search_suggestion.js`, I guess in parallel with that effort, we can try to convert that file to TypeScript to make refactoring within it easier... which is blocked only by https://github.com/zulip/zulip/pull/28542.

evykassirer commented:
Oh yeah sorry I didn't update -- I was looking into this more in the past week and there are a handful of things that need changes. One is I need to support group DMs, which will be a longish process and seems required for this PR being able to get merged. The other is I need to tweak things in search_suggestions, as you alluded to. I'll update you when those things are more cleaned up.

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -0,0 +1,49 @@
+    const items = pill_widget.items();
+    const search_strings = items.map((item) => item.display_value);
+    return search_strings.join(" ");
+}
```
Yeah definitely! Thanks for noticing

evykassirer commented on web/styles/search.css:
```diff
@@ -187,6 +187,10 @@
+        /* Override .pill-container */
+        flex-wrap: nowrap;
+        padding: 0;
+        border: none;
```
Would `/* Override styles for .pill-container that aren't relevant for the search pill container` be sufficient? There were some styles set up to be shared for all pill containers, but the search box has its own set of styles it needs to have instead.

timabbott replied:
Sure

evykassirer commented on web/src/search.js:
```diff
@@ -58,25 +67,47 @@ export function initialize({on_narrow_search}) {
-            return suggestions.strings;
+            const existing_pill_strings = new Set(
+                search_pill_widget.items().map((item) => item.display_value),
+            );
```
(not relevant anymore due to passing existing pills to the function finding search suggestions)

evykassirer commented:
Still very much in draft state, but pushing some of the stuff I was working on most recently (and rebased)

evykassirer commented:
This most recent push will include functional group DM pills! I think everything mostly works? But it's still loosely tested (so likely missing edge case issues) and I want to do a lot of cleaning up code still.

I'm pushing to
(1) More easily view the code I wrote and separate it into prep commits
(2) Allow for testing if Alya wants to play around with it and provide early feedback

There are some issues with the colors of the pills, that I just asked about on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20pills/near/1792273).

evykassirer commented on web/src/input_pill.ts:
```diff
@@ -148,29 +151,32 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
             }
+            let pill_html;
+            if (item.type === "search_user") {
+                pill_html = render_search_user_pill(item);
```
It feels a bit weird to have this custom logic here, but that's sort of a consequence of sharing a lot of code with `input_pill` (that's encapsulated in here) and also wanting to do a lot of custom things.

timabbott replied:
It's not obvious to me that `render_input_pill` couldn't be adapted to handle this, but I think this is fine and it's best to try to do further refactoring around that interface in a follow-up PR, that perhaps tries to use object inheritance.

evykassirer replied:
I'll add that to a list of followups :)

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -32,6 +56,36 @@ export function create_pills($pill_container: JQuery): SearchPillWidget {
+        type: "search_user",
+        operator,
+        // TODO(evy): make this "search_string" instead of "display_value"
+        display_value: search_string,
```
It would be nice if we just call this `search_string` instead of `display_value`, but `display_value` is needed to hook into the generic input pill logic.

timabbott replied:
The implementation here seems reasonable. Let's memorialize this thought as a `TODO` comment in the `SearchUserPill` type definition (or wherever feels most appropriate).

evykassirer replied:
`SearchUserPill` type definition makes sense to me, put it there :)

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -752,42 +754,74 @@ function get_operator_suggestions(last: NarrowTerm): Suggestion[] {
 
+// One search suggestion can include multiple search terms, based on what's
+// in the search bar.
+type SuggestionLine = Suggestion[];
```
Would it be good to rename `Suggestion` to something like `SuggestionTerm` to more clearly differentiate from `SuggestionLine`?

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -752,42 +754,74 @@ function get_operator_suggestions(last: NarrowTerm): Suggestion[] {
+function suggestion_search_string(suggestion_line: SuggestionLine): string {
+    const search_strings = [];
+    for (const suggestion of suggestion_line) {
+        if (suggestion.search_string !== "") {
```
One way we can get empty search strings here is when creating an `Attacher` with a `base` that's an empty term. I decided it was cleaner to filter it out here than try to remove it somewhere else. Does that seem reasonable?

evykassirer commented on web/src/input_pill.ts:
```diff
@@ -430,6 +470,15 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
         // select the next pill (or input).
         store.$parent.on("click", ".exit", function (e) {
+            // TODO(evy) this really should live in search land, not here, but
+            // I don't know a good way to do that.
```
I even had to put the click handler here because `funcs` isn't public to other files.

timabbott replied:
I think this is also probably fine to live here for now, just add a comment as with my suggestion on the other part of the commit, that both notes a TODO and also explains what `$user_pill_container.length` means clearly to someone who doesn't know anything about search.

Maybe best would be inside the `$user_pill_container.length` block, a comment like>

> // The user-pill-container container class is used exclusively for group-DM search pills, where multiple user pills sit inside a larger pill. The exit icons in those individual user pills should remove just that pill, not the outer pill.
> // TODO: Figure out how to move this code into `search_pill.ts`.

evykassirer replied:
Cool, thanks, did that!

evykassirer commented on web/src/input_pill.ts:
```diff
@@ -243,6 +244,45 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
         },
 
+        // This is only used for search, and maybe shouldn't live here?
+        removeUserPill(user_container: HTMLElement, user_id: number) {
```
I'd love some help with this commit. It's functional, but the code feels messy in several ways and I think I should be implementing it differently but I'm not sure what would be better.

Here are the concerns I have:

(1) The way this code reads feels a bit clunky (several `for` loops finding indexes of things) but I'm mirroring this off of `removePill` and I think it might be needed to be able to reach into the `store` to delete the pill?

(2) I feel weird putting this code in `input_pill` because it's very specific to search and not generic at all, but it does seem to need `store` to be able to edit the pill content. Would it be better to put this code elsewhere and have the `store` less encapsulated? Or is it better to just have this function here?

(3) ` const user_pill_container: InputPillItem<SearchUserPill> = store.pills[container_idx].item;` is having type error issues because the `store` treats the pills as template types. This feels related to the issues above, but I'm not sure what a good way of addressing it is.

timabbott replied:
This seems readable enough to me. I'd consider just replacing the comment just above `removeUserPill` with something like:

> // TODO: This function is only used for the search input supporting multiple user pills within an individual top-level pill. Ideally, we'd encapsulate it in a subclass used only for search so that this code can be part of `search_pill.ts`.

evykassirer replied:
Okay, I'll add that comment and leave it there for now. Thanks!

evykassirer commented:
(encountered a bug while rebasing, don't test yet!)

evykassirer commented:
Mostly working except I just realized after pushing that deleting pills still isn't fully functional. I'll update here when I've fixed that!

evykassirer commented:
@alya should be ready for you now!

alya commented:
Thanks! I'm not 100% sure what the right types of feedback are, but here's what I noticed:

1. We shouldn't replace ` ` with `+`, now that the text is inside pills:

![Screenshot 2024-05-09 at 13 22 07](https://github.com/zulip/zulip/assets/2090066/9ebea9c2-9329-4af3-a37f-6a82c5084673)

2. I think it would look better to put a ` ` after the `:`, e.g. here:
![Screenshot 2024-05-09 at 13 19 06](https://github.com/zulip/zulip/assets/2090066/ffebaa08-9e13-4cd7-ae48-7a764c5c5a18)


4.  Searching for an invalid `dm:` gives an error:
![Screenshot 2024-05-09 at 13 24 53](https://github.com/zulip/zulip/assets/2090066/3a0d4ed9-95a2-4dd4-a76b-b78207912803)
![Screenshot 2024-05-09 at 13 25 12](https://github.com/zulip/zulip/assets/2090066/4f64d82d-57ef-4089-a6fd-faac65e92bf7)

alya commented:
4. Something weird happens when you click `x` on a pill inside a dm-with pill with multiple users

![Screenshot 2024-05-09 at 13 25 52](https://github.com/zulip/zulip/assets/2090066/1035be70-f3b2-458e-8359-1869f62b90e4)

5. When you click the `x` on a pill, it gets highlighted in green just before it disappears. This feels unnecessary/distracting; I don't know if that's something to fix separately.

alya commented:
6. I'm not sure if it's a question for @terpimost / discussed elsewhere, but I fount it very hard to see the pill boundaries in dark theme.
![Screenshot 2024-05-09 at 13 37 15](https://github.com/zulip/zulip/assets/2090066/202ba492-5f92-4399-ad14-feba27870428)

alya commented:
7. I think the cursor is always too high? Elsewhere, it's centered on the text, and it doesn't seem to be here. I believe this is separate from the specific instance noted in the PR description.

I think that's all I've got right now!

evykassirer commented:
I think 4/5/6 will all be addressed by [these design questions I asked about](https://chat.zulip.org/#narrow/stream/101-design/topic/search.20pills/near/1794287) last week. But more specifically:

(4) Does it close the pill? X-ing a pill should still work, but yes it does highlight the group-user-pill before closing. I didn't bother fixing it looking weird yet, because I asked for further direction on how to style that. Depending on what selecting the pill ends up looking like, I might want to prevent the pill from getting selected before it closes.

(5) Oh that's just the same logic as (4), which again might actually be fine if we change the highlight to outline. You can see this behavior already exists for the PM pills, so I wouldn't worry about it specifically for search here.

I'll look into making the changes for (1) and (2), and (3) is definitely a bug thanks for catching that! I'll look into that too. I'll leave (7) for when I investigate the other cursor issue.

This was very helpful, thanks! Glad that it seems mostly working as expected, for someone else as well.

evykassirer commented:
Pushing a rebased version with the color changes Vlad suggested. But I'll make a new PR for those color changes shortly so we can test and integrate them for non-search places.

evykassirer commented on web/src/input_pill.ts:
```diff
@@ -163,6 +164,7 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
                     has_image,
                     deactivated: item.deactivated,
                     should_add_guest_user_indicator: item.should_add_guest_user_indicator,
+                    is_user: opts.type === "user",
```
`item.type`, I don't know how that got through

evykassirer commented on web/styles/search.css:
```diff
@@ -225,8 +225,7 @@
             min-width: fit-content;
-            /* TODO colors -- dark mode? other pills? */
-            background-color: hsl(201deg 49% 93%);
+            background-color: var(--color-background-input-pill);
```
don't need these

evykassirer commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -280,6 +297,10 @@ export class Typeahead<ItemType extends string | object> {
+        // It's possible that we got here from pressing enter with nothing highlighted.
+        if (this.allowNoHighlight && val === undefined) {
+            return this.hide();
+        }
```
I'd rather we just not get to `select` at all in this case, but there are a lot of calls to `select` and I don't want to miss any.

evykassirer commented on web/src/message_view_header.ts:
```diff
@@ -149,7 +149,6 @@ function build_message_view_header(filter: Filter | undefined): void {
     // message_view_header on a template where it's never used
     if (filter && !filter.is_common_narrow()) {
         search.open_search_bar_and_close_narrow_description();
-        search.set_search_bar_text(narrow_state.search_string());
```
This will already happen from `reset_searchbox` which called through `open_search_bar_and_close_narrow_description` (not in every case, but in the cases that we want).

evykassirer commented on web/src/search.ts:
```diff
@@ -151,17 +184,10 @@ export function initialize({on_narrow_search}: {on_narrow_search: OnNarrowSearch
-
-                // Pill is already added during keydown event of input pills.
-                narrow_or_search_for_term(get_search_bar_text(), {on_narrow_search});
-                $search_query_box.trigger("blur");
```
This already happens most of the time (when we want it to) in `narrow_or_search_for_term`

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -41,6 +94,16 @@ export function set_search_bar_contents(
+            const user_emails = term.operand.split(",");
+            const users = user_emails.map((email) => {
+                const user = people.get_by_email(email);
+                assert(user !== undefined);
```
Handling of invalid users is in an upcoming commit.

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -768,7 +802,24 @@ class Attacher {
+                suggestion_line = [suggestion];
+            } else {
+                // When we add a user to a user group, we
+                // replace the last pill.
```
This feels a bit janky, but I think it would take a significant refactor to have more structured data here and I've been trying to keep the changes simpler where I can.

alya commented:
Hm, how come the font size is so much smaller in the search pills than in the compose box? And why is the `x` bigger in the search pill?

![Screenshot 2024-06-13 at 12 15 04](https://github.com/zulip/zulip/assets/2090066/dccb4453-3f44-4989-a668-2f084c7f4c6e)

evykassirer commented:
I have the CSS rules matching the DM rules now, does this look right to you?

<img width="638" alt="image" src="https://github.com/zulip/zulip/assets/5634097/91d1b2b7-0d28-43cb-8118-7368f2b6413f">

alya commented:
Looks good based on the screenshot!

evykassirer commented:
rebased to remove conflicts

timabbott commented:
I noticed it doesn't support the new `is:followed` filter yet.

![image](https://github.com/zulip/zulip/assets/2746074/26293913-1ee3-4541-b691-d38871553a3c)

timabbott commented:
None of those seem so broken that we can't test this, and it doesn't look like the remaining TODOs are so disruptive that we can't test, so I'm test-deploying this on chat.zulip.org to try to shake out more potential problems.

evykassirer commented:
> Using `/` to open search doesn't seem to apply keyboard focus properly.

I wasn't able to reproduce something that looked like an issue to me. Can you explain more what happens and what you're expecting?

timabbott commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -471,6 +494,13 @@ export class Typeahead<ItemType extends string | object> {
+        // whatever's already in the search box.
+        if (!this.requireHighlight && $active.length && !$next.length) {
+            return;
+        }
```
Can you explain the theory behind why returning with no action is the right answer for `prev/next` when no highlight is required? It looks like this implementation has `down` when at the bottom do nothing, not wrap around (the default) or switch to nothing being highlighted.

evykassirer replied:
There is an action -- we switch to nothing being highlighted. See `this.$menu.find(".active").removeClass("active");` above.

If you press down twice when at the bottom, the first time it removes the highlight and second time puts it at the top of the list.

evykassirer commented on web/src/search.ts:
```diff
@@ -51,12 +51,22 @@ function narrow_or_search_for_term({on_narrow_search}: {on_narrow_search: OnNarr
+        get_search_bar_text(),
+    ]
+        .filter((s) => s !== "")
+        .join(" ");
```
This is kind of ugly -- is there a nicer way to write this?

timabbott replied:
Is it intended to what it looks like it does, which is concatenate those strings with a `" "` between if they're both nonempty? Maybe something like `[x, y].join(" ").trim()` might be cleaner?

evykassirer replied:
Ah perfect, using trim sounds much better, thanks!

evykassirer commented on web/src/search.ts:
```diff
@@ -51,12 +51,22 @@ function narrow_or_search_for_term({on_narrow_search}: {on_narrow_search: OnNarr
     }
     const terms = Filter.parse(search_query);
+    // Reset the search bar, in case some of these terms
+    // haven't been pillified yet.
```
We need to do this now because `convert_to_pill_on_enter` doesn't. But it's better because here we can do it our custom way.

timabbott replied:
This seems fine, just explain what we're doing and why, with any relevant context on how `convert_to_pill_on_enter` is related to this (Was it already run, but didn't clear the text? Or are we doing this instead of it? Etc.)

evykassirer replied:
Yes exactly it's because `convert_to_pill_on_enter`, I'll add a comment

evykassirer commented:
I split out some changes to the typeahead and input_pill options, making them commits before the bulk of the other changes. I'm not actually sure if this makes the PR easier to review, but I've heard requests for this kind of separation in the past so hopefully it helps! I also added some more descriptions in the commit messages.

timabbott commented on web/src/search_suggestion.ts:
```diff
@@ -982,7 +987,10 @@ export function get_search_result(query_from_pills: string, query_from_text: str
+    } else if (
+        all_search_terms.length > 0 &&
+        !(page_params.is_spectator && ["has", "is"].includes(last.operator))
+    ) {
```
What's the story behind the new `is_spectator` check here?

evykassirer replied:
I don't completely remember, but I agree it's strange. It's on my list to look into, but I want to start with [this conversation about invalid is](https://chat.zulip.org/#narrow/stream/101-design/topic/showing.20invalid.20is.20pills/near/1839707) since depending on how that goes, this might become irrelevant and removed.

timabbott commented on web/src/search_pill.ts:
```diff
@@ -89,45 +99,86 @@ function append_user_pill(
+        default:
+            return true;
+    }
+}
```
This seems to not be exhaustive; the `streams:` operator should limit what's valid here.

evykassirer replied:
Oh oops, I thought `streams` couldn't reach here and only `channel`, but I found a way where it can, thanks!

evykassirer replied:
> Might be worth putting this in as Filter.is_valid_search_pill, defined just after Filter.parse ; you could do that as a prep PR that defines the function, with tests, while unused, just to get logic integrated separately from the main PR.

I've set up that as a prep commit and made a PR for it here #30678

timabbott commented on web/src/search_pill.ts:
```diff
@@ -89,45 +99,86 @@ function append_user_pill(
+        case "in":
+            return ["home", "all"].includes(term.operand);
+        case "id":
+            return message_store.get(Number.parseInt(term.operand, 10)) !== undefined;
```
the `id:` and `near:` operators should behave identically, and just validate that they got an integer parameter. Checking `message_store` is just looking at a local cache, and thus incorrect validation -- we can't know which message IDs are valid locally. (And both are intended to allow non-existent message IDs; it's important for `near` functionally to allow stuff like `near:1` to find the start of a view)

timabbott commented on web/src/search_pill.ts:
```diff
@@ -89,45 +99,86 @@ function append_user_pill(
+        if (user_pill_operators.has(term.operator) && term.operand !== "") {
+            const users = term.operand.split(",").map((email) => {
+                const user = people.get_by_email(email);
+                assert(user !== undefined);
```
Can you add a string on this assert statement for why we believe it to always be valid? I think the reason is we're reading out of an already-completed pill?  `assert(test, "Email should be for a valid user for already-completed pill")` or something.

I think that it might be possible to have that fail if a user's email address changed and we didn't live-update the search query? This should be fixed when we change the pills to store `user_id` rather than email, which I hope to be a near-term follow-up to this PR.

evykassirer replied:
It's because `is_valid_pill` checks this directly above.

evykassirer replied:
I'll add a code comment, but I'm not sure if we need a string? I think we could even switch this to a `!` instead of an assert -- I'll do that for now.

timabbott commented on web/src/search_pill.ts:
```diff
@@ -89,45 +99,86 @@ function append_user_pill(
+    }
+    set_search_bar_text(search_bar_text_strings.join("  "));
+    if (invalid_inputs.length) {
+        $("#search_query").addClass("shake");
```
Why is there a double-space used in this join?

evykassirer replied:
That's a mistake! Thanks

timabbott commented on web/src/input_pill.ts:
```diff
@@ -168,8 +168,25 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+                if (item.type === "search") {
+                    display_value = display_value.replaceAll("+", " ");
+                    display_value = display_value.replace(":", ": ");
+                }
```
This hack is fine, since it's well-contained. Might be good to have a `TODO:` at the start of the "Ideally, "` paragraph.

timabbott commented on web/src/input_pill.ts:
```diff
@@ -255,6 +256,59 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
+            // TODO(evy): It doesn't seem easy with typescript here to have
+            // the item be assigned to the SearchUserPill type, even though it
+            // has a `type: "search_user"`. Not sure how easily fixable this is
+            // without a big refactor.
```
I think this type of detail is something we can use an `as` unchecked type assertion for and follow-up on, though it wouldn't be a bad idea to pair with Anders on it, since he's been fixing other issues with pill types recently.

evykassirer replied:
Cool I'll do that for now and also reach out to @andersk.

evykassirer commented:
Most recent push changes how we deal with invalid is/has https://chat.zulip.org/#narrow/stream/101-design/topic/showing.20invalid.20is.20pills

evykassirer commented:
No remaining issues that I'm aware of

alya commented:
Testing this PR, we're letting long names go to far to the right (overlapping with the `x`):

![Screenshot 2024-07-03 at 15 28 07](https://github.com/zulip/zulip/assets/2090066/f559e71f-dd10-4838-8ed7-63954cf608f0)

Here the name gets abbreviated, and but the status emoji is too far right:

![Screenshot 2024-07-03 at 15 26 47](https://github.com/zulip/zulip/assets/2090066/e0157436-9fa8-47a5-86f0-24dd8781cc52)

timabbott commented:
I reordered the commits a bit for clarity, and squashed a couple that were pure bug fixes to an original commit (mainly `fixup: Only focus search input on init and update, not page load.` and the close button for direct-message-group pills). I think merging with the commit history as written here will probably make life a little easier for debugging issues/regressions caused by changes in one of the later commits, so I reworked the commit messages to not claim we're going to squash them.

Merged, thanks for all the work on this @evykassirer! We should coordinate follow-ups in the design thread. The main thing that's notable to me as a likely engineering-implementation follow-up is that I think we need to rework the core model for search pills to store a canonical value for these pills that is separate from the display value. The goal would be to clean up:
- Stream IDs, not stream names, should be what's stored.
- The `+` format for strings with spaces in them should not only be present in the code for generating a URL and parsing one; we should be storing the actual name in the pill object, not the URL-encoded name.
- Fully migrate the web app to sending channels to the server entirely using IDs, not channel names, in the `message_fetch` collection of APIs. Currently there's this hack, `handle_operators_supporting_id_based_api`, where we convert channel names into IDs at the API boundary, but that could be completely avoided if we didn't store them as strings in the search system in the first place.

I don't think it's required that such refactoring complete for 9.0, but we might as well work on it while this is all fresh.
